### PR TITLE
[retry] Allow stopping retry with a custom option: WithMaxDuration

### PR DESCRIPTION
**Pull request by Soulou**

### CODE REVIEW

**Comment by Soulou** [comment]
It allows me to write this, 

```
	retrier := retry.New(
		retry.WithWaitDuration(client.containerStateCheckInterval),
		retry.WithTimeout(client.containerDeathTimeout),
	)
	err := retrier.Do(ctx, func(ctx context.Context) error {
```
instead of this:

```
	retrier := retry.New(
		retry.WithWaitDuration(client.containerStateCheckInterval),
		retry.WithMaxAttempts(int(client.containerDeathTimeout/client.containerStateCheckInterval)+10),
	)
	ctx, cancel := context.WithTimeout(ctx, client.containerDeathTimeout)
	defer cancel()

	err := retrier.Do(ctx, func(ctx context.Context) error {
```

Which is according to me much more elegant

File: retry/retry.go [hunk 1]
```diff
@@ -54,11 +81,10 @@ func (r Retryer) Do(ctx context.Context, method Retryable) error {
 		timer := time.NewTimer(r.waitDuration)
 		select {
 		case <-timer.C:
+		case <-timeoutCtx.Done():
+			return timeoutCtx.Err()
```
Comment (johnsudaar) [inline retry/retry.go]: You should specialize the context errors here.

Comment (Soulou) [reply retry/retry.go]: Done

File: retry/retry.go [hunk 2]
```diff
@@ -2,9 +2,27 @@ package retry
...
 )
 
+type RetryErrScope string
+
+const (
+	MaxDurationScope RetryErrScope = "max-duration"
+	ContextScope     RetryErrScope = "context"
+)
+
+type RetryErr struct {
```
Comment (EtienneM) [inline retry/retry.go]: Following this golang doc (https://github.com/golang/go/wiki/Errors#naming), you should name this:
```suggestion
type RetryError struct {
```

Comment (Soulou) [reply retry/retry.go]: Done

**Pull request merged**